### PR TITLE
[SPARK-13332][SQL] Decimal datatype support for  SQL pow

### PR DESCRIPTION
**Pull request by yucai**
Decimal datatype support for SQL pow
- when base is Decimal and exponent is integer(Byte, Short, Int), calculate and return as Decimal
- otherwise, calculate and return as Double

https://issues.apache.org/jira/browse/SPARK-13332

### CODE REVIEW

**Comment by yucai** [comment]
@adrian-wang could you help review? Much thanks!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala [hunk 1]
```diff
@@ -103,8 +103,7 @@ class MathFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
       }
     } else {
       domain.foreach { case (v1, v2) =>
-        checkEvaluation(c(Literal(v1), Literal(v2)), f(v1 + 0.0, v2 + 0.0), EmptyRow)
-        checkEvaluation(c(Literal(v2), Literal(v1)), f(v2 + 0.0, v1 + 0.0), EmptyRow)
+        checkEvaluation(c(Literal(v1), Literal(v2)), f(v1, v2), EmptyRow)
```
Comment (adrian-wang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: keep the test of `f(v2, v1)`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala [hunk 2]
```diff
@@ -351,6 +350,20 @@ class MathFunctionsSuite extends SparkFunSuite with ExpressionEvalHelper {
   }
 
   test("pow") {
+    testBinary(Pow, (d: Decimal, n: Byte) => d.pow(n),
+      (-5 to 5).map(v => (Decimal(v * 1.0), v.toByte)))
```
Comment (adrian-wang) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/MathFunctionsSuite.scala]: maybe `v.toDouble` is better

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java [hunk 3]
```diff
@@ -170,6 +170,7 @@ public void write(int ordinal, double value) {
   }
 
   public void write(int ordinal, Decimal input, int precision, int scale) {
+    input = input.clone();
```
Comment (adrian-wang) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: Better add a comment that explains why we need to clone before write.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala [hunk 4]
```diff
@@ -523,11 +523,45 @@ case class Atan2(left: Expression, right: Expression)
...
+      case (dt: DecimalType, ShortType) =>
+        input1.asInstanceOf[Decimal].pow(input2.asInstanceOf[Short])
+      case (dt: DecimalType, IntegerType) =>
+        input1.asInstanceOf[Decimal].pow(input2.asInstanceOf[Int])
+      case (dt: DecimalType, FloatType) =>
+        math.pow(input1.asInstanceOf[Decimal].toDouble, input2.asInstanceOf[Float])
+      case (dt: DecimalType, DoubleType) =>
+        math.pow(input1.asInstanceOf[Decimal].toDouble, input2.asInstanceOf[Double])
+      case (dt1: DecimalType, dt2: DecimalType) =>
+        math.pow(input1.asInstanceOf[Decimal].toDouble, input2.asInstanceOf[Decimal].toDouble)
```
Comment (adrian-wang) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/mathExpressions.scala]: Shall we cast the result of `math.pow` back to `DecimalType` for these three cases?

**Comment by rxin** [comment]
I think it'd be a lot simpler if we create a separate Pow for Decimal, and handle byte/short/etc to integer in type coercion, rather than in the PowDecimal class.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java (hunk 3 above)
Comment (rxin) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: Why is this necessary? Seems like a really bad idea.

Comment (adrian-wang) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: Here we'll call `changePrecision` on `input` here, which would affect the orignal data. I agree that this is a bad idea, maybe we need to propose a separate pr to work around this.

Comment (yucai) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/codegen/UnsafeRowWriter.java]: As Adrian mentioned, we need a copy of input, otherwise `changePrecision` would change the original input.
In our case, this means `catalystValue`(expected value) would be changed when `checkEvalutionWithUnsafeProjection` is invoked, and then all tests after checkEvalutionWithUnsafeProjection will fail.
```
  protected def checkEvaluation(
      expression: => Expression, expected: Any, inputRow: InternalRow = EmptyRow): Unit = {
    val catalystValue = CatalystTypeConverters.convertToCatalyst(expected)
    checkEvaluationWithoutCodegen(expression, catalystValue, inputRow)
    checkEvaluationWithGeneratedMutableProjection(expression, catalystValue, inputRow)
    if (GenerateUnsafeProjection.canSupport(expression.dataType)) {
      checkEvalutionWithUnsafeProjection(expression, catalystValue, inputRow)
    }
    checkEvaluationWithOptimization(expression, catalystValue, inputRow)
  }
```
Does it make sense? Any suggestion is great helpful.

**Comment by yucai** [comment]
OK, let me try this implementation.

**Comment by yucai** [comment]
@rxin I tried your suggestion like creating PowDecimal for Decimal specially like below:
 
```
case class PowDecimal(left: Expression, right: Expression)
  extends BinaryMathExpression(math.pow, "POWER") {
  override def inputTypes: Seq[AbstractDataType] = Seq(DecimalType, IntegerType)
  ...
}

case class Pow(left: Expression, right: Expression)
  extends BinaryMathExpression(math.pow, "POWER") {
  ...
} 
``` 

But one concern is when "select pow(cast(2 as decimal(5,2)), 3)", how to make the "PowDecimal" node created? The current path will create "Pow" node anyway.

So we think of maybe we can still put Decimal processing in "Pow", but byte/short/etc. to integer in type coercion, like below:

```
case class Pow(left: Expression, right: Expression)
  extends BinaryMathExpression(math.pow, "POWER") {
  override def inputTypes: Seq[AbstractDataType] = Seq(NumericType, NumericType)

  override def dataType: DataType = (left.dataType, right.dataType) match {
    case (dt: DecimalType, ByteType | ShortType | IntegerType) => dt
    case _ => DoubleType
  }
  protected override def nullSafeEval(input1: Any, input2: Any): Any =
    (left.dataType, right.dataType) match {
      case (dt: DecimalType, _) => input1.asInstanceOf[Decimal].pow(input2.asInstanceOf[Int])
      case _ => math.pow(input1.asInstanceOf[Double], input2.asInstanceOf[Double])
    }
  override def genCode(ctx: CodegenContext, ev: ExprCode): String = ...
}
  
In HiveTypeCoercion:
```
  object PowCoercion extends Rule[LogicalPlan] {
    def apply(plan: LogicalPlan): LogicalPlan = plan resolveExpressions {
      case e if !e.childrenResolved => e
      case e @ Pow(left, right) =>
        (left.dataType, right.dataType) match {
          case (dt: DecimalType, IntegerType) => e
          case (DoubleType, DoubleType) => e
          case (dt: DecimalType, ByteType | ShortType) =>
            Pow(left, Cast(right, IntegerType))
          case _ => Pow(Cast(left, DoubleType), Cast(right, DoubleType))
        }
    }
  }
```
How do you think this way?

**Comment by rxin** [comment]
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one. We can also continue the discussion on the JIRA ticket.
